File: app/src/main/java/com/donald/musictheoryapp/Question/Question.java
package com.donald.musictheoryapp.Question;

import androidx.annotation.NonNull;

import org.json.JSONException;
import org.json.JSONObject;
import org.xmlpull.v1.XmlPullParserException;

import java.io.File;
import java.io.IOException;
import java.util.Arrays;

public abstract class Question
{
    // TODO: MAKE INTEGER NULLABLE
    public static final int MULTIPLE_CHOICE = 0;
    public static final int TEXT_INPUT = 1;
    public static final int CHECK_BOX = 2;
    public static final int TRUTH = 3;
    public static final int INTERVAL_INPUT = 4;

    public interface QuestionVisitor
    {
        void visit(MultipleChoiceQuestion question);
        void visit(TextInputQuestion question);
        void visit(TruthQuestion question);
        void visit(CheckBoxQuestion question);
        void visit(IntervalInputQuestion question);
    }

    public interface Answer
    {
        boolean correct();
    }

    public int number;
    public QuestionGroup group;
    public Description[] descriptions;

    public abstract int points();

    public abstract void acceptVisitor(QuestionVisitor visitor);

    public static Question fromJSON(JSONObject object, QuestionGroup group) throws JSONException, IOException, XmlPullParserException
    {
        Question question;
        switch (object.getInt("type"))
            {
                case MULTIPLE_CHOICE:
                    question = MultipleChoiceQuestion.fromJSON(object, group);
                    break;
                case TEXT_INPUT:
                    question = TextInputQuestion.fromJSON(object, group);
                    break;
                case CHECK_BOX:
                    question = CheckBoxQuestion.fromJSON(object, group);
                    break;
                case TRUTH:
                    question = TruthQuestion.fromJSON(object, group);
                    break;
                case INTERVAL_INPUT:
                    question = IntervalInputQuestion.fromJSON(object, group);
                    break;
                default:
                    throw new JSONException("Question type " + object.getInt("type") + " unrecognized.");
            };
        return question;
    }
}

File: app/src/main/java/com/donald/musictheoryapp/question/Question.kt
package com.donald.musictheoryapp.question

import com.donald.musictheoryapp.util.getQuestionType
import kotlin.Throws
import org.xmlpull.v1.XmlPullParserException
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.io.IOException

abstract class Question(
    val number: Int,
    val descriptions: Array<Description>,
    val inputHint: String?
) {

    abstract val points: Int
    abstract val maxPoints: Int

    abstract fun acceptVisitor(visitor: QuestionVisitor)

    open fun registerImages(imageArray: JSONArray) {
        for (description in descriptions) {
            if (description.type == Description.Type.IMAGE) {
                imageArray.put(description.content)
            }
        }
    }

    inline fun visit(
        crossinline whenMultipleChoice: (MultipleChoiceQuestion) -> Unit,
        crossinline whenTextInput: (TextInputQuestion) -> Unit,
        crossinline whenTruth: (TruthQuestion) -> Unit,
        crossinline whenCheckBox: (CheckBoxQuestion) -> Unit,
        crossinline whenIntervalInput: (IntervalInputQuestion) ->  Unit
    ) {
        acceptVisitor(
            object : QuestionVisitor {
                override fun visit(question: MultipleChoiceQuestion) = whenMultipleChoice(question)
                override fun visit(question: TextInputQuestion) = whenTextInput(question)
                override fun visit(question: TruthQuestion) = whenTruth(question)
                override fun visit(question: CheckBoxQuestion) = whenCheckBox(question)
                override fun visit(question: IntervalInputQuestion) = whenIntervalInput(question)
            }
        )
    }

    @Throws(JSONException::class)
    protected abstract fun toPartialJson(): JSONObject
    @Throws(JSONException::class)
    fun toJson(): JSONObject {
        val jsonObject = toPartialJson()
        // 1. number
        jsonObject.put("number", number)

        // 2. descriptions
        val currentArray = JSONArray()
        for (description in descriptions) currentArray.put(description.toJson())
        jsonObject.put("descriptions", currentArray)

        // 3. panel hint
        val inputHint = inputHint
        if (inputHint == null) jsonObject.put("input_hint", JSONObject.NULL)
        else jsonObject.put("input_hint", inputHint)
        return jsonObject
    }

    companion object {

        @JvmStatic
        @Throws(JSONException::class, IOException::class, XmlPullParserException::class)
        fun fromJson(jsonObject: JSONObject): Question {
            val question: Question = when (jsonObject.getQuestionType()) {
                Type.MULTIPLE_CHOICE -> MultipleChoiceQuestion.fromJson(jsonObject)
                Type.TEXT_INPUT -> TextInputQuestion.fromJson(jsonObject)
                Type.CHECK_BOX -> CheckBoxQuestion.fromJson(jsonObject)
                Type.TRUTH -> TruthQuestion.fromJson(jsonObject)
                Type.INTERVAL_INPUT -> IntervalInputQuestion.fromJson(jsonObject)
            }
            return question
        }

    }

    enum class Type { MULTIPLE_CHOICE, TEXT_INPUT, CHECK_BOX, TRUTH, INTERVAL_INPUT }

    interface QuestionVisitor {

        fun visit(question: MultipleChoiceQuestion)
        fun visit(question: TextInputQuestion)
        fun visit(question: TruthQuestion)
        fun visit(question: CheckBoxQuestion)
        fun visit(question: IntervalInputQuestion)

    }

    interface Answer {

        val correct: Boolean

    }

}